libsql-wal/src/shared_wal.rs

```diff
@@ -0,0 +1,212 @@
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Instant;
+
+use arc_swap::ArcSwap;
+use crossbeam::deque::Injector;
+use crossbeam::sync::Unparker;
+use libsql_sys::wal::PageHeaders;
+use parking_lot::{Mutex, MutexGuard};
+
+use crate::error::{Error, Result};
+use crate::fs::file::FileExt;
+use crate::fs::FileSystem;
+use crate::name::NamespaceName;
+use crate::registry::WalRegistry;
+use crate::segment::current::CurrentSegment;
+use crate::transaction::{ReadTransaction, Savepoint, Transaction, WriteTransaction};
+
+#[derive(Default)]
+pub struct WalLock {
+    tx_id: Mutex<Option<u64>>,
+    reserved: Mutex<Option<u64>>,
```

Comment:
```suggestion
    /// When a writer is popped from the write queue, its write transaction may not be reading from the most recent
    /// snapshot. In this case, we return `SQLITE_BUSY_SNAPHSOT` to the caller. If no reads were performed
    /// with that transaction before upgrading, then the caller will call us back immediately after re-acquiring
    /// a read mark.
    /// Without the reserved slot, the writer would be re-enqueued, a writer before it would be inserted,
    /// and we'd find ourselves in the initial situation. Instead, we use the reserved slot to bypass the queue when the 
    /// writer tried to re-acquire the write lock.
    reserved: Mutex<Option<u64>>,
```